### PR TITLE
Use Chariott's new service discovery endpoint

### LINKED ISSUE #8: [PBI] Use new service discovery mechanism with Chariott.
## Description

Chariott has a new service discovery endpoint. This will update the Pub Sub service to utilize it.

## Acceptance criteria

- [x] Same functionality provided with the old service.
- [x] Uses new service discovery endpoint from chariott.

### PR DESCRIPTION
**Pull request by devkelley**
Closes #8 

## Motivation and Context

Chariott's new service discovery mechanism is easier to interact with and is the direction that the service is heading in. As Agemo is close to Chariott, updating to the new mechanism is important.

## Description

This PR utilizes the new service discovery endpoint, which allows providers to register with the service without needing to stand up a gRPC server for interaction. This PR makes the following changes:
- Removes old Chariott registration process, which included a register and announce loop and a gRPC server endpoint.
- Simplifies folder structure since less code is needed for Chariott communication
- Adds helper functions to handle interacting with Chariott through the generated gRPC client.

### CODE REVIEW

File: samples/common/src/chariott_helper.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/// Helper function for initiating the Chariott client connection. Retries on failure.
+/// Returns a gRPC client.
+///
+/// # Arguments
+///
+/// * `chariott_url` - The url for Chariott.
+/// * `retry_interval_secs` - The interval to wait before retrying the connection.
+pub async fn connect_to_chariott_with_retry(
+    chariott_url: &str,
+    retry_interval_secs: u64,
```
Comment (ashbeitz) [inline samples/common/src/chariott_helper.rs:26]: u64 is a big number, would something smaller suffice?

**Review by ashbeitz — COMMENTED** [review]
Please take a look at my comments.

File: pub-sub-service/src/connectors/chariott_connector.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                if status.code() == Code::Unavailable {
+                    reason = String::from("No chariott service found");
+                } else {
+                    reason = format!("Chariott request failed with '{status:?}'");
+                };
+                None
+            }
+        }
+        .or_else(|| {
+            let secs = 5;
```
[the remark this answers is not in the document]
Comment (devkelley) [reply pub-sub-service/src/connectors/chariott_connector.rs:55]: I will have this updated with this issue #3. Updated issue include constants in the description.

File: pub-sub-service/src/main.rs [hunk 3]
```diff
@@ -46,47 +36,14 @@ pub mod topic_manager;
 /// Endpoint for the messaging broker.
 const BROKER: &str = "mqtt://localhost:1883";
```
[the remark this answers is not in the document]
Comment (devkelley) [reply pub-sub-service/src/main.rs:37]: Will be taken care of with #3

File: samples/chariott-publisher/src/main.rs [hunk 4]
```diff
@@ -5,99 +5,75 @@
...
-    let provider_url = Url::parse(&chariott_provider_client.get_provider_url()).unwrap();
+    Ok(())
+}
 
-    Ok(ChariottProvider::new(provider_url))
+/// Calls Chariott to get Pub Sub Service endpoint.
+///
+/// # Arguments
+///
+/// * `chariott_url` - The Chariott url.
```
[the remark this answers is not in the document]
Comment (devkelley) [reply samples/chariott-publisher/src/main.rs:61]: Ah missed one, thanks

File: samples/common/src/lib.rs [hunk 5]
```diff
@@ -2,13 +2,14 @@
...
+pub mod chariott_helper;
 pub mod data_generator;
 pub mod pub_sub_service_helper;
 pub mod publisher_helper;
 pub mod subscriber_helper;
 pub mod topic_store;
 
 pub mod constants {
-    pub const CHARIOTT_ENDPOINT: &str = "http://0.0.0.0:4243";
+    pub const CHARIOTT_ENDPOINT: &str = "http://0.0.0.0:50000";
```
[the remark this answers is not in the document]
Comment (devkelley) [reply samples/common/src/lib.rs:13]: Will address with #3

File: pub-sub-service/src/connectors/chariott_connector.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+use proto::{
+    pubsub,
+    service_registry::v1::service_registry_client::ServiceRegistryClient,
+    service_registry::v1::{RegisterRequest, ServiceMetadata},
+};
+
+type ChariottClient = ServiceRegistryClient<Channel>;
+
+/// Object that contains the necessary information for identifying a specific service.
+pub struct ServiceIdentifiers {
```
[the remark this answers is not in the document]
Comment (devkelley) [reply pub-sub-service/src/connectors/chariott_connector.rs]: I am attempting to keep this consistent to what Chariott uses. I was informed that it is actually ServiceIdentifier (singular) so have updated it to that.

File: pub-sub-service/src/connectors/chariott_connector.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+//! Set of helper functions for interacting with Chariott through the generated gRPC client.
+
+use std::{thread, time::Duration};
+
+use log::{info, warn};
+use tonic::{transport::Channel, Code, Request, Status};
+
```
[the remark this answers is not in the document]
Comment (devkelley) [reply pub-sub-service/src/connectors/chariott_connector.rs:10]: removed

File: samples/common/src/chariott_helper.rs (hunk 1 above)
Comment (devkelley) [reply samples/common/src/chariott_helper.rs:26]: Good question. So the wait mechanism I am using uses [Duration](https://doc.rust-lang.org/std/time/struct.Duration.html) and the seconds are of type u64. I could probably cast it, but this was the default. Let me know your thoughts.

Comment (ashbeitz) [reply samples/common/src/chariott_helper.rs:26]: Leave it as is then.

File: pub-sub-service/src/connectors/chariott_connector.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashbeitz) [reply pub-sub-service/src/connectors/chariott_connector.rs]: It's fine then.

**Review by ashbeitz — APPROVED** [review]
Approved.

**Pull request merged**